core/src/main/scala/kafka/log/AbstractIndex.scala

```diff
@@ -34,16 +34,37 @@ import scala.math.ceil
 /**
  * The abstract index class which holds entry format agnostic methods.
  *
- * @param file The index file
+ * @param _file The index file
  * @param baseOffset the base offset of the segment that this index is corresponding to.
  * @param maxIndexSize The maximum index size in bytes.
  */
-abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Long,
+abstract class AbstractIndex[K, V](@volatile var _file: File, val baseOffset: Long,
                                    val maxIndexSize: Int = -1, val writable: Boolean) extends Closeable with Logging {
 
+  def file: File = {
+    if (!initialized) {
+      inLock(lock) {
+        if (!initialized) {
+          initializeMmapAndLength
+        }
+      }
+    }
+    _file
+  }
+
+  def fileName: String = {
+    _file.getName
+  }
+
+  def file_=(f: File) {
+    _file = f
+  }
```

Comment:
nit: I generally prefer listing member variables before methods

Reply:
Since this is more of a code style question and @hzxa21 may not be familiar with that, I will provide some comment here. So the question here is that should we put listing member variables together at the beginning of the class, or should we put each member variable close to its getter/setter.

It seems that the existing code style in Kafka puts the member variables together at the beginning of the class. This is the case for `highWatermarkMetadata`, `_logStartOffset` and `_lastCaughtUpTimeMs` in `Replica.scala`. The benefit of this approach is that developer can easily see all states (i.e. member variable) of the class at the beginning of the file instead of having to look through the entire file to understand the states. On the other hand, even if we don't put member variable close to its getter/setter, developer can still easily know the member variable associated with each getter/setter by just looking at the method name etc.

Reply:
vals/vars should be at the top since the order matters. It's hard to reason about it if they are scattered all over the file. Methods (`defs`) don't have this issue.

Reply:
Fixed.